### PR TITLE
[python-package] Allow to pass Arrow array as groups

### PR DESCRIPTION
**Pull request by borchero**
# Motivation

This is part of a series of PRs towards #6022 and depends on #6164.

For a legible diff, see https://github.com/borchero/LightGBM/compare/arrow-support-weights...arrow-support-groups.

### CODE REVIEW

File: src/io/metadata.cpp [hunk 1]
```diff
@@ -507,30 +507,34 @@ void Metadata::InsertWeights(const label_t* weights, data_size_t start_index, da
   // CUDA is handled after all insertions are complete
 }
 
-void Metadata::SetQuery(const data_size_t* query, data_size_t len) {
+template <typename It>
+void Metadata::SetQueriesFromIterator(It first, It last) {
   std::lock_guard<std::mutex> lock(mutex_);
-  // save to nullptr
-  if (query == nullptr || len == 0) {
+  // Clear weights on empty input
```
Comment (jameslamb) [inline src/io/metadata.cpp:513]:
```suggestion
  // Clear query boundaries on empty input
```

File: src/io/metadata.cpp [hunk 2]
```diff
@@ -507,30 +507,34 @@ void Metadata::InsertWeights(const label_t* weights, data_size_t start_index, da
...
     query_boundaries_.clear();
     num_queries_ = 0;
     return;
   }
+
   data_size_t sum = 0;
-  #pragma omp parallel for num_threads(OMP_NUM_THREADS()) schedule(static) reduction(+:sum)
-  for (data_size_t i = 0; i < len; ++i) {
-    sum += query[i];
+  #pragma omp parallel for schedule(static) reduction(+:sum)
```
Comment (jameslamb) [inline src/io/metadata.cpp:521]:
```suggestion
  #pragma omp parallel for num_threads(OMP_NUM_THREADS()) schedule(static) reduction(+:sum)
```

I'm not sure why this pragma was modified, but please revert that change. It's important for fine-grained control of multithreading that every `omp parallel` in the project uses the LightGBM-controlled configuration `OMP_NUM_THREADS()` to determine the degree of parallelism.

See #6135 and the things it links to for some background.

File: src/io/metadata.cpp [hunk 3]
```diff
@@ -507,30 +507,34 @@ void Metadata::InsertWeights(const label_t* weights, data_size_t start_index, da
...
+  for (data_size_t i = 0; i < last - first; ++i) {
+    sum += first[i];
   }
   if (num_data_ != sum) {
-    Log::Fatal("Sum of query counts is not same with #data");
+    Log::Fatal("Sum of query counts differs from the length of #data");
```
Comment (jameslamb) [inline src/io/metadata.cpp:526]:
```suggestion
    Log::Fatal("Sum of query counts (%i) differs from the length of #data (%i)", num_data_, sum);
```

While you're touching this line anyway, would you consider modifying this error message such that it includes the actual sizes? That would be helpful in debugging why this error was encountered.

File: tests/python_package_test/test_arrow.py [hunk 4]
```diff
@@ -183,3 +186,32 @@ def test_dataset_construct_weights(array_type: Any, weight_data: Any, arrow_type
...
+        pa.int16(),
+        pa.int32(),
+        pa.int64(),
+        pa.uint8(),
+        pa.uint16(),
+        pa.uint32(),
+        pa.uint64(),
+    ],
+)
+def test_dataset_construct_groups(array_type: Any, group_data: Any, arrow_type: Any):
```
Comment (jameslamb) [inline tests/python_package_test/test_arrow.py:210]:
```suggestion
def test_dataset_construct_groups(array_type, group_data, arrow_type):
```

Please don't add type hints on `pytest` test functions in this project. Especially when they're all set to `Any`, they just make the code a little harder to read without providing any documentation or help to type-checkers.

Can you please remove them? And please also remove them from the test cases introduced in the previous Arrow PRs? Sorry I missed that before.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Changes look good to me, just left a few small requests.

I've also removed the `WIP: ` from the title since this is now ready to review.

File: tests/python_package_test/test_arrow.py [hunk 5]
```diff
@@ -38,13 +37,13 @@ def generate_dummy_arrow_table() -> pa.Table:
     return pa.Table.from_arrays([col1, col2], names=["a", "b"])
 
 
-def generate_random_arrow_table(num_columns: int, num_datapoints: int, seed: int) -> pa.Table:
+def generate_random_arrow_table(num_columns, num_datapoints, seed) -> pa.Table:
```
Comment (jameslamb) [inline tests/python_package_test/test_arrow.py:40]: @borchero I might not have been clear in https://github.com/microsoft/LightGBM/pull/6166#discussion_r1392709299.

Please preserve the type hints for arguments and returns on utility functions referenced from tests... they're helpful documentation of how data is flowing through the tests.

I just was asking that you remove type hints for `pytest` test cases (any function whose name begins with `test_`).

Comment (borchero) [reply tests/python_package_test/test_arrow.py:40]: Whoops, sure :smile:

File: tests/python_package_test/test_arrow.py [hunk 6]
```diff
@@ -175,11 +178,40 @@ def test_dataset_construct_weights_none():
...
 
     expected = np.array([3, 0.7, 1.5, 0.5, 0.1], dtype=np.float32)
     np_assert_array_equal(expected, dataset.get_weight(), strict=True)
+
+
+# -------------------------------------------- GROUPS ------------------------------------------- #
+
+
+@pytest.mark.parametrize(
+    ["array_type", "group_data"], [(pa.array, [2, 3]), (pa.chunked_array, [[2], [3]])]
```
Comment (jameslamb) [inline tests/python_package_test/test_arrow.py:195]: Could you please add test cases covering the case where a chunked array with some empty chunks is given?

At least these two:

* the first chunk of the chunked array is empty
* one or more other non-first chunk of the chunked array is empty

And then could you:

* put up a separate PR adding such test cases for the other `Dataset` components you've already added Arrow support for
* add such test cases in the remaining Arrow PRs

Sorry I didn't think of this earlier. It does seem to me that it's possible to have empty chunks in a chunked array, e.g.

```python
import pyarrow as pa

arr = pa.chunked_array(
    [[2, 3, 4], []],
    type=pa.int16()
)

pa.compute.sum(arr)
# <pyarrow.Int64Scalar: 9>
```

And I can imagine situations where such arrays could end up being passed into LightGBM. I'm thinking use cases similar to how Dask Arrays are commonly created by concatenating the results of multiple separate function calls, some of which may be empty, e.g.:

* reading chunks from multiple files (and maybe applying row filters while reading them)
* executing multiple database queries
* paging over results from a REST API
* streaming over chunks returned from a database (e.g. [like in `snowflake-connector-python`](https://github.com/snowflakedb/snowflake-connector-python/blob/9b6b0a63887e472e7c05365e5db896bf4d818db1/src/snowflake/connector/result_set.py#L113))

**Review by jameslamb — CHANGES_REQUESTED** [review]
These changes look good to me! I'm +1 on supporting Arrow arrays and chunked arrays for `group`, and good with the proposed changes to the Python package and added tests.

I just left one more suggestion, around also testing for chunked arrays with empty chunks.

@borchero could you please update this branch to latest `master` to get the newest changes from there?

@shiyu1994 @guolinke could you help with another review?

File: tests/python_package_test/test_arrow.py (hunk 6 above)
Comment (borchero) [reply tests/python_package_test/test_arrow.py:195]: Ha! I didn't know that this was possible, will add tests!

**Review by guolinke — APPROVED** [review]
C++ part LGTM!

**Comment by borchero** [comment]
Here we finally go @jameslamb :smile: let me know if this kind of test suffices and I will put up another PR to add these kinds of tests for the dataset, labels, and weights.

**Review by jameslamb — APPROVED** [review]
New tests look great to me, thanks! I'll merge this once it builds.

**Comment by jameslamb** [comment]
Thank you! Let's go to the next one 😁
